lib/ws2_manager.js

```diff
@@ -0,0 +1,363 @@
+   */
+  getNumSockets () {
+    return this._sockets.length
+  }
+
+  /**
+   * @param {number} i
+   * @return {Object} state
+   */
+  getSocket (i) {
+    return this._sockets[i]
+  }
+
+  /**
+   * Returns an object which can be logged to inspect the socket pool
+   */
+  getSocketInfo () {
+    return this._sockets.map(s => ({
+      nChannels: WS2Manager.getDataChannelCount(s)
+    }))
+  }
+
+  /**
+   * Authenticates all existing & future sockets with the provided credentials.
+   * Does nothing if an apiKey/apiSecret pair are already known.
+   *
+   * @param {Object} args
+   * @param {Object?} args.apiKey - saved if not already provided
+   * @param {Object?} args.apiSecret - saved if not already provided
+   * @param {Object?} args.calc - default 0
+   * @param {Object?} args.dms - dead man switch, active 4
+   */
+  auth ({ apiKey, apiSecret, calc = 0, dms = 0 } = {}) {
+    if (this._socketArgs.apiKey || this._socketArgs.apiSecret) {
+      debug('error: auth credentials already provided! refusing auth')
+      return
+    }
+
+    this._socketArgs.apiKey = apiKey
+    this._socketArgs.apiSecret = apiSecret
+    this._authArgs = { calc, dms }
+
+    this._sockets.forEach(s => {
+      if (!s.ws.isAuthenticated()) {
+        s.ws.setAuthCredentials(apiKey, apiSecret)
+        s.ws.auth({ calc, dms })
+      }
+    })
+  }
+
+  /**
+   * Creates a new socket/state instance and adds it to the internal pool. Binds
+   * event listeners to forward via our own event emitter, and to manage pending
+   * subs/unsubs.
+   *
+   * @return {Object} state
+   */
+  openSocket () {
+    const { apiKey, apiSecret } = this._socketArgs
+    const { calc, dms } = this._authArgs
+    const ws = new WSv2(this._socketArgs)
+    const wsState = {
+      pendingSubscriptions: [],
+      pendingUnsubscriptions: [],
+      ws
+    }
+
+    ws.once('open', () => debug('socket connection opened'))
+    ws.on('open', () => this.emit('open', ws))
+    ws.on('message', (msg = {}) => this.emit('message', msg, ws))
+    ws.on('error', (error) => this.emit('error', error, ws))
+    ws.on('auth', () => this.emit('auth', ws))
+    ws.on('close', () => this.emit('close', ws))
+    ws.on('subscribed', (msg = {}) => {
+      const i = wsState.pendingSubscriptions.find(sub => {
+        const fv = _pick(msg, Object.keys(sub[1]))
+
+        return (
+          (sub[0] === msg.channel) &&
+          _isEqual(fv, sub[1])
+        )
+      })
+
+      if (i === -1) {
+        debug('error removing pending sub: %j', msg)
+        return
+      }
+
+      wsState.pendingSubscriptions.splice(i, 1)
+      this.emit('subscribed', msg)
+    })
+
+    ws.on('unsubscribed', (msg = {}) => {
+      const { chanId } = msg
+      const i = wsState.pendingUnsubscriptions.find(cid => (
+        cid === chanId
+      ))
+
+      if (i === -1) {
+        debug('error removing pending unsub: %j', msg)
+        return
+      }
+
+      wsState.pendingUnsubscriptions.splice(i, 1)
```

Comment:
hows does it work? i thought find returns the value, not the key: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find

Reply:
can you add a test for this?

Reply:
Nice catch, you are correct! It should be `findIndex`, will fix